tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,244 @@
+      continue
+
+    curated = row[utils.CURATED_ALTERNATIVES_COL].split(";")
+    palm_alts = row[utils.ALTERNATIVES_COL].split(";")
+
+    # Pair name and description with very high score.
+    if name != desc:
+      training_examples.append(
+          InputExample(texts=[name, desc], label=VERY_HIGH_MATCH_SCORE))
+
+    # If there are more descriptions, pair them with the name as well.
+    if len(descriptions) > 1:
+      for i in range(1, len(descriptions)):
+        if descriptions[i]:
+          training_examples.append(
+              InputExample(texts=[name, descriptions[i]],
+                           label=HIGH_MATCH_SCORE))
+
+    for c in curated:
+      # All all curated alternatives as pairs with the description.
+      if c:
+        training_examples.append(
+            InputExample(texts=[desc, c], label=HIGH_MATCH_SCORE))
+
+    for p in palm_alts:
+      if p:
+        # High match score but since these are auto-generated, keep the score lower
+        # than for the manual/curated cases.
+        training_examples.append(
+            InputExample(texts=[name, p], label=MEDIUM_HIGH_MATCH_SCORE))
+
+  # Now add the manual sentence pairs (which are assumed to have the scores).
+  for _, row in df_sentence_pairs.iterrows():
+    assert "sentence_1" in row
+    assert "sentence_2" in row
+    assert "score" in row
+    training_examples.append(
+        InputExample(texts=[row["sentence_1"], row["sentence_2"]],
+                     label=row["score"]))
+
+  # Ready to fine-tune.
+  train_dataloader = DataLoader(training_examples,
+                                shuffle=True,
+                                batch_size=BATCH_SIZE)
+  t = time.time()
+  model.fit(train_objectives=[(train_dataloader,
+                               losses.CosineSimilarityLoss(model=model))],
+            epochs=NUM_EPOCHS,
+            warmup_steps=NUM_WARMUP_STEPS)
+  t = time.time() - t
+  print(f"Time taken = {t}")
+  return model
+
+
+def main(_):
+
+  assert FLAGS.model_name_v2 and FLAGS.bucket_name_v2
+  assert FLAGS.embeddings_size in ['small', 'medium']
+
+  assert os.path.exists(os.path.join('data'))
+
+  autogen_input_filepattern = f'{FLAGS.autogen_input_basedir}/{FLAGS.embeddings_size}/*.csv'
+
+  gs = gspread.oauth()
+  sc = storage.Client()
+  bucket = sc.bucket(FLAGS.bucket_name_v2)
+
+  df_svs = _alternatives(autogen_input_filepattern,
+                         FLAGS.alternatives_filepattern)
+  df_sentence_pairs = pd.read_csv(FLAGS.sentence_pairs_filepath).fillna("")
+  model = SentenceTransformer(FLAGS.model_name_v2)
+  model = fine_tune_model(model, df_svs, df_sentence_pairs)
+
+  ctx = utils.Context(gs=gs, model=model, bucket=bucket, tmp='/tmp')
+
+  gcs_model_folder = _make_gcs_model_folder(FLAGS.model_name_v2)
+  gcs_tmp_out_path = os.path.join(ctx.tmp, gcs_model_folder)
+
+  print(f"Saving locally to {gcs_tmp_out_path}")
+  model.save(gcs_tmp_out_path)
+
+
+#   # Finally, upload to the NL model server's GCS bucket
```

Comment:
Are the commented code still useful?

Reply:
Oooops! Yes that code was supposed to be uncommented :)